### PR TITLE
Add support for triples of positions in volume

### LINKED ISSUE #222: scan volume and other functions like that should be able to accept blocks or triples as args like other functions.

### PR DESCRIPTION
**Pull request by Ghoulboy78**
Fixes #222

### CODE REVIEW

**Comment by Ghoulboy78** [comment]
how's that?

File: src/main/java/carpet/script/CarpetExpression.java [hunk 1]
```diff
@@ -2096,21 +2096,29 @@ private void API_IteratingOverAreasOfBlocks()
...
-            int minz = min(zi, zj);
-            int maxx = max(xi, xj);
-            int maxy = max(yi, yj);
-            int maxz = max(zi, zj);
-            LazyValue expr = lv.get(6);
+        this.expr.addLazyFunction("volume", -1, (c, t, lv) ->
+        {
+            CarpetContext cc = (CarpetContext)c;
+
+            if(lv.size()!=3 & lv.size()!=5 & lv.size()!=7) throw new InternalExpressionException("'volume' takes 3, 5 or 7 arguments.");
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2103]: you should rely on .findIn to report argument errors

File: src/main/java/carpet/script/CarpetExpression.java [hunk 2]
```diff
@@ -2096,21 +2096,29 @@ private void API_IteratingOverAreasOfBlocks()
...
+            int x2 = pos2.getX();
+            int y2 = pos2.getY();
+            int z2 = pos2.getZ();
+            int minx = min(x1, x2);
+            int miny = min(y1, y2);
+            int minz = min(z1, z2);
+            int maxx = max(x1, x2);
+            int maxy = max(y1, y2);
+            int maxz = max(z1, z2);
+            LazyValue expr = lv.get(lv.size()-1);
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2121]: likewise, you should use pos2.offset to assert you have a correct argument, not size-1

File: src/main/java/carpet/script/CarpetExpression.java [hunk 3]
```diff
@@ -2195,50 +2206,61 @@ private void API_IteratingOverAreasOfBlocks()
...
+                cz = cpos.getZ();
+
+                sminx = 1;
+                sminy = 1;
+                sminz = 1;
+                smaxx = 1;
+                smaxy = 1;
+                smaxz = 1;
+
+                if(lv.size()<3){
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2223]: use comparison between lv.size() and locator.offset here instead

File: src/main/java/carpet/script/CarpetExpression.java [hunk 4]
```diff
@@ -2195,50 +2206,61 @@ private void API_IteratingOverAreasOfBlocks()
...
-                else // size == 9
-                {
-                    sminx = (int) ((NumericValue) lv.get(3).evalValue(c)).getLong();
-                    sminy = (int) ((NumericValue) lv.get(4).evalValue(c)).getLong();
-                    sminz = (int) ((NumericValue) lv.get(5).evalValue(c)).getLong();
-                    smaxx = (int)((NumericValue) lv.get(6).evalValue(c)).getLong();
-                    smaxy = (int)((NumericValue) lv.get(7).evalValue(c)).getLong();
-                    smaxz = (int)((NumericValue) lv.get(8).evalValue(c)).getLong();
+
+                    if(lv.size()<6){
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2235]: ditto

File: src/main/java/carpet/script/CarpetExpression.java [hunk 5]
```diff
@@ -2298,22 +2320,29 @@ public String getString()
...
             {
-                throw new InternalExpressionException("'diamond' region should be specified with 3 to 5 coordinates");
+                throw new InternalExpressionException("'diamond' region should be specified with 3 to 5 arguments");
             }
 
+            int blockposarg=2;//Checks if the first arg is a block pos triple
+
+            if(lv.get(0).evalValue(c) instanceof NumericValue)blockposarg=0;
+
+            BlockPos cpos=BlockArgument.findIn((CarpetContext)c,lv,0).block.getPos();
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2332]: findIn should raise appropriate exceptions then use offset to determine the remaining arguments, and raise exceptions as needed

**Comment by Ghoulboy78** [comment]
just a sec, workin on th other one rn, that one's easier

File: src/main/java/carpet/script/CarpetExpression.java [hunk 6]
```diff
@@ -2193,52 +2198,60 @@ private void API_IteratingOverAreasOfBlocks()
...
-                    smaxx = (int)((NumericValue) lv.get(6).evalValue(c)).getLong();
-                    smaxy = (int)((NumericValue) lv.get(7).evalValue(c)).getLong();
-                    smaxz = (int)((NumericValue) lv.get(8).evalValue(c)).getLong();
-                }
-            }
-            catch (ClassCastException exc)
-            {
-                throw new InternalExpressionException("Attempted to pass a non-number to rect");
+
+            } catch (IndexOutOfBoundsException i){//Checking if there is only 1 pos
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2229]: using try/catch for regular logic is very detrimental to the performance. Use try-catches only to catch errors that would flip the script anyways.

Use if (lv.size() > locator.offset).... to continue with optional parameters

File: src/main/java/carpet/script/CarpetExpression.java [hunk 7]
```diff
@@ -2326,15 +2339,15 @@ public String getString()
...
-                    width = (int) ((NumericValue) lv.get(3).evalValue(c)).getLong();
+                    width = (int) ((NumericValue) lv.get(lv.size()-1).evalValue(c)).getLong();
                     height = 0;
                 }
-                else // size == 5
+                else // size == 3 + cposLocator.offset
                 {
-                    width = (int) ((NumericValue) lv.get(3).evalValue(c)).getLong();
-                    height = (int) ((NumericValue) lv.get(4).evalValue(c)).getLong();
+                    width = (int) ((NumericValue) lv.get(lv.size()-2).evalValue(c)).getLong();
```
Comment (gnembon) [inline src/main/java/carpet/script/CarpetExpression.java:2349]: again - don't count list from the back, just use offsets to count from the front.

**Comment by Ghoulboy78** [comment]
done.

**Pull request merged**